Use undef for uninitialized bytes in constants

erikdesjardins opened the pull request:
Fixes #83657

This generates good code when the const is fully uninit, e.g.

```rust
#[no_mangle]
pub const fn fully_uninit() -> MaybeUninit<[u8; 10]> {
    const M: MaybeUninit<[u8; 10]> = MaybeUninit::uninit();
    M
}
```
->
```asm
foo:
	ret
```

as you would expect.

There is no improvement, however, when it's partially uninit, e.g.

```rust
pub struct PartiallyUninit {
    x: u64,
    y: MaybeUninit<[u8; 10]>
}

#[no_mangle]
pub const fn partially_uninit() -> PartiallyUninit {
    const X: PartiallyUninit = PartiallyUninit { x: 0xdeadbeefcafe, y: MaybeUninit::uninit() };
    X
}
```
->
```asm
partially_uninit:
	mov	rax, rdi
	mov	rcx, qword ptr [rip + .L__unnamed_1+16]
	mov	qword ptr [rdi + 16], rcx
	movups	xmm0, xmmword ptr [rip + .L__unnamed_1]
	movups	xmmword ptr [rdi], xmm0
	ret

.L__unnamed_1:
	.asciz	"\376\312\357\276\255\336\000"
	.zero	16
	.size	.L__unnamed_1, 24
```
which copies a bunch of zeros in place of the undef bytes, the same as before this change.

The constant is undef where it should be, though (see the added test) so this seems to be a missed optimization in LLVM.

rust-highfive commented:
r? @estebank

(rust-highfive has picked a reviewer for you, use r? to override)

oli-obk commented:
r? @oli-obk

oli-obk commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -867,6 +874,50 @@ impl InitMask {
+        }
+
+        let is_init = self.init_mask.get(self.start);
+        let end_of_chunk = (self.start.bytes()..self.end.bytes())
+            .map(Size::from_bytes)
```
Maybe we should impl the `Step` trait for `Size`, but that doesn't have to happen in this PR

oli-obk commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -867,6 +874,50 @@ impl InitMask {
+        let is_init = self.init_mask.get(self.start);
+        let end_of_chunk = (self.start.bytes()..self.end.bytes())
+            .map(Size::from_bytes)
+            .find(|&i| self.init_mask.get(i) != is_init)
```
I think we could improve on this by checking entire blocks instead of going through the entries in the block, but again, doesn't have to happen in this PR (please leave a FIXME comment for that though).

erikdesjardins replied:
Dealing with the {start,end} of the {search range,initialized chunk} being unaligned with block boundaries seems subtle (and very branchy in a naive implementation) so I'll just copy the fixme from the other place this was done

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit be66c2e551c177c9607f257ccadefb0c64820645 has been approved by `oli-obk`

<!-- @bors r=oli-obk be66c2e551c177c9607f257ccadefb0c64820645 -->
<!-- homu: {"type":"Approved","sha":"be66c2e551c177c9607f257ccadefb0c64820645","approver":"oli-obk"} -->

erikdesjardins commented:
I don't think this should be rolled up, it could affect perf or expose existing UB (because previously reading uninitialized bytes from constants at runtime would "work")

RalfJung commented:
@bors rollup=never

@Dylan-DPC might be better to re-roll that rollup... sorry for this.

bors commented:
:hourglass: Testing commit be66c2e551c177c9607f257ccadefb0c64820645 with merge 24f957fcbe9f9432f0c74cc707a5f008268f4984...
<!-- homu: {"type":"BuildStarted","head_sha":"be66c2e551c177c9607f257ccadefb0c64820645","merge_sha":"24f957fcbe9f9432f0c74cc707a5f008268f4984"} -->

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/2257082426) [(plain)](https://github.com/rust-lang-ci/rust/commit/24f957fcbe9f9432f0c74cc707a5f008268f4984/checks/2257082426/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
status: exit status: 2
command: "make"
stdout:
------------------------------------------
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/wasm-panic-small/wasm-panic-small:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/wasm-panic-small/wasm-panic-small -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/wasm-panic-small/wasm-panic-small  foo.rs -C lto -O --target wasm32-unknown-unknown --cfg a
wc -c < /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/wasm-panic-small/wasm-panic-small/foo.wasm
1139
[ "`wc -c < /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/wasm-panic-small/wasm-panic-small/foo.wasm`" -lt "1024" ]
------------------------------------------
stderr:
------------------------------------------
------------------------------------------
make: *** [Makefile:8: all] Error 1
------------------------------------------




failures:
    [run-make] run-make/wasm-panic-small

test result: FAILED. 16 passed; 1 failed; 6 ignored; 0 measured; 0 filtered out; finished in 2.47s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/wasm32-unknown-unknown/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--src-base" "/checkout/src/test/run-make" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make" "--stage-id" "stage2-wasm32-unknown-unknown" "--suite" "run-make" "--mode" "run-make" "--target" "wasm32-unknown-unknown" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--nodejs" "/node-v14.4.0-linux-x64/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/wasm32-unknown-unknown/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--llvm-version" "12.0.0-rust-1.53.0-nightly" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xray" "--llvm-bin-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin" "--cc" "clang" "--cxx" "clang" "--cflags" "-ffunction-sections -fdata-sections -fPIC --target=wasm32-unknown-unknown" "--ar" "ar" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --host= --target wasm32-unknown-unknown src/test/run-make src/test/ui src/test/mir-opt src/test/codegen-units library/core
Build completed unsuccessfully in 0:17:22
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2257188976)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/2257188976","builder_name":"checks-actions"} -->

erikdesjardins commented:
> ```
> wc -c < /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/wasm-panic-small/wasm-panic-small/foo.wasm
> 1139
> [ "`wc -c < /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/wasm-panic-small/wasm-panic-small/foo.wasm`" -lt "1024" ]
> ```

When building this test locally, it's never below 1024 bytes (which is strange), but there is a regression: at HEAD~2 foo.wasm is 6k, and with these changes it's 140k (!). It seems to be defeating some optimization in LLVM, so we're no longer able to eliminate a bunch of formatting code. I'm trying to figure out why.

It may be worth a perf run to see if this change as-is causes widespread regressions (I suspect it will), or if that wasm test is just fragile for some reason

nagisa commented:
Now that you mention this, In my memory llvm would fail to put constants
with unit them into bss sections in the past. Does wasm have something
similar?

On Tue, 6 Apr 2021, 06:34 erikdesjardins, ***@***.***> wrote:

> wc -c < /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/wasm-panic-small/wasm-panic-small/foo.wasm
> 1139
> [ "`wc -c < /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/wasm-panic-small/wasm-panic-small/foo.wasm`" -lt "1024" ]
>
> When building this test locally, it's never below 1024 bytes (which is
> strange), but there is a regression: at HEAD~2 foo.wasm is 6k, and with
> these changes it's 140k (!). It seems to be defeating some optimization in
> LLVM, so we're no longer able to eliminate a bunch of formatting code. I'm
> trying to figure out why.
>
> It may be worth a perf run to see if this change as-is causes widespread
> regressions (I suspect it will), or if that wasm test is just fragile for
> some reason
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/83698#issuecomment-813796976>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFFZUWSLBZABFD6GSH6RF3THJ6MNANCNFSM42DX4N6A>
> .
>

RalfJung commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit be66c2e551c177c9607f257ccadefb0c64820645 with merge e3b9e122cd5f8e8e4c9903fb4458f274e462758b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"be66c2e551c177c9607f257ccadefb0c64820645","merge_sha":"e3b9e122cd5f8e8e4c9903fb4458f274e462758b"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2276596659)
Build commit: e3b9e122cd5f8e8e4c9903fb4458f274e462758b (`e3b9e122cd5f8e8e4c9903fb4458f274e462758b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2276596659"},"merge_sha":"e3b9e122cd5f8e8e4c9903fb4458f274e462758b"} -->

rust-timer commented:
Queued e3b9e122cd5f8e8e4c9903fb4458f274e462758b with parent a6e7a5aa5dc842ddb64607040aefb6ec33e22b59, future [comparison URL](https://perf.rust-lang.org/compare.html?start=a6e7a5aa5dc842ddb64607040aefb6ec33e22b59&end=e3b9e122cd5f8e8e4c9903fb4458f274e462758b).

rust-timer commented:
Finished benchmarking try commit (e3b9e122cd5f8e8e4c9903fb4458f274e462758b): [comparison url](https://perf.rust-lang.org/compare.html?start=a6e7a5aa5dc842ddb64607040aefb6ec33e22b59&end=e3b9e122cd5f8e8e4c9903fb4458f274e462758b).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

RalfJung commented:
Wow those are some mixed results.^^ I am not sure how to interpret them...

bors commented:
:hourglass: Testing commit be66c2e551c177c9607f257ccadefb0c64820645 with merge 7a29d6b30b70fe1bf16868a2e0d0e3c9c7bfd34e...
<!-- homu: {"type":"BuildStarted","head_sha":"be66c2e551c177c9607f257ccadefb0c64820645","merge_sha":"7a29d6b30b70fe1bf16868a2e0d0e3c9c7bfd34e"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/2295581185) [(plain)](https://github.com/rust-lang-ci/rust/commit/7a29d6b30b70fe1bf16868a2e0d0e3c9c7bfd34e/checks/2295581185/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 82  424M   82  348M    0     0  8413k      0  0:00:51  0:00:42  0:00:09 11.1M
 84  424M   84  359M    0     0  8471k      0  0:00:51  0:00:43  0:00:08 11.1M
 87  424M   87  370M    0     0  8523k      0  0:00:50  0:00:44  0:00:06 10.8M
 88  424M   88  374M    0     0  8556k      0  0:00:50  0:00:44  0:00:06 10.8M
curl: (56) OpenSSL SSL_read: Connection reset by peer, errno 54
clang+llvm-10.0.0-x86_64-apple-darwin/bin/opt: Lzma library error:  No progress is possible
tar: Error exit delayed from previous errors.
##[error]Process completed with exit code 1.
[command]/usr/local/bin/git version
git version 2.31.1
[command]/usr/local/bin/git config --local --name-only --get-regexp core\.sshCommand
[command]/usr/local/bin/git submodule foreach --recursive git config --local --name-only --get-regexp 'core\.sshCommand' && git config --local --unset-all 'core.sshCommand' || :
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2295617127)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/2295617127","builder_name":"checks-actions"} -->

tmiasko commented:
The tuple-stress & ucd benchmarks contain large arrays with padding inside, for example `static BIG_ARRAY_OF_TUPLES: [(i32, (f64, f64, f64)); 0xffff] = [ ... ]`, `pub static UCD_NUMS: &'static [(i64,u8)] = &[ ... ]`, which are now more expensive to codegen.

erikdesjardins commented:
I'm still not sure why this PR makes that .wasm file huge.

It is not reproducible for non-wasm targets.
While the .wasm binary format does have something like bss (memory is zero by default, so you can omit zero initializers), the optimized LLVM IR is much larger as well (compare `test-opt-before.ll` [~43kB] and `test-opt-after.ll` [~1000kB] below), so I don't think that's the culprit.

What I wanted to do was get the unoptimized LLVM IR before and after this change, differing only by undef bytes, hopefully reproduce it by feeding that to `opt`, then reduce it to the minimal IR that causes a size blowup.
However, it seems to require LTO, which makes things difficult. `-Cno-prepopulate-passes` has no effect (?) when `-Clto` is passed, so I don't know how get the unoptimized IR.
With just `-Cno-prepopulate-passes` and no LTO, the before and after files only differ by `{ [0 x i8] }` being replaced by `{}` in one constant (compare `test-noopt-before.ll` and `test-noopt-after.ll` below), which "surely" can't be enough to cause this, and passing both `test-noopt-*.ll` files to `opt -O3` produces identical output as you'd expect.

Referenced files can be found here: https://gist.github.com/erikdesjardins/b73f23cf6bd515ec4a00855c1a2edfd1

oli-obk commented:
Two things:

1. We could only do the undef marking for constants that already contain relocations (and thus are already expensive to codegen) or are very small (less than a kb or sth)
2. We could skip doing this for wasm (and open an issue about this topic)

Aaron1011 commented:
Somehow, bors thinks that this PR is approved
@bors r-

erikdesjardins commented:
@oli-obk now allocs > 1kB will be either fully undef (if possible) or fully initialized. Requiring relocations would be too restrictive, it would prevent this from fixing #83657. Large, fully-undef allocs are supported so that `MaybeUninit<very large array>` will still be optimized.

Also disabled this on wasm targets.

nikic commented:
Is there any reason to believe that this is a wasm-specific issue? I'd have assumed it just manifests there, because that target happens to be used for the LTO code size test.

erikdesjardins commented:
It doesn't happen when compiling the same code for other targets.

e.g. (stage1base is 58bdb08947f5b3c18a2fbafc5cf36af7b5677d83; stage1 is 3e576e3, this branch with the wasm special case removed)

test.rs

```rust
#![crate_type = "cdylib"]

#[no_mangle]
pub fn foo() {
    panic!("test");
}
```

```
$ rustc +stage1base -C lto -O test.rs --target wasm32-unknown-unknown && wc -c test.wasm
6873 test.wasm
$ rustc +stage1 -C lto -O test.rs --target wasm32-unknown-unknown && wc -c test.wasm
139666 test.wasm

$ rustc +stage1base -C lto -O test.rs && wc -c libtest.so
1534160 libtest.so
$ rustc +stage1 -C lto -O test.rs && wc -c libtest.so
1536632 libtest.so
```

There is a 2k size regression on x86_64-unknown-linux-gnu, but it's nothing like wasm.

It's possible that it happens on all targets, but only cfg(wasm) code in std currently hits it.

erikdesjardins commented:
In that wasm test, without this change, we manage to remove all of the arguments from `rust_panic_with_hook`, along with a lot of panic formatting/location code.
This doesn't happen when we generate partially-undef consts, which causes the size regression, for some reason (perhaps due to nearby changes to the type of `std::panicking::HOOK`, from `{ [8 x i8] }` to `{ [4 x i8], [4 x i8] }`).
To see this, diff preundef-specific.ll and postundef-specific.ll from this gist: https://gist.github.com/erikdesjardins/0e5b2f26c1c246b451b489f68577855a.

So it isn't wasm specific--it only shows up on wasm because the default wasm panic handler just traps without using any of the panic info, which allows the further simplification.

I don't have the energy to reduce this further. I tried to extract `_ZN3std9panicking20rust_panic_with_hook17hc745a002b86303d0E` (with its dependencies `declare`d) and pass it to `opt -O3`, but I couldn't get the same optimization to happen.

I removed the wasm special case and changed this PR to only generate consts that are fully-init or fully-undef, which seems to fix the issue. Generating partially-undef consts is now behind `-Zpartially-uninit-const-threshold=1234`.

RalfJung commented on compiler/rustc_session/src/options.rs:
```diff
@@ -1068,6 +1068,9 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
         "parse only; do not compile, assemble, or link (default: no)"),
+    partially_uninit_const_threshold: Option<usize> = (None, parse_opt_number, [TRACKED],
+        "allow generating const initializers with mixed init/uninit bytes, \
+        and set the maximum size at which this is allowed (default: never)"),
```
Maximum size of what -- of the const allocation, or of one of the (un)init ranges? I tried understanding this from the code but failed...

RalfJung commented on compiler/rustc_codegen_llvm/src/consts.rs:
```diff
@@ -11,19 +11,70 @@ use rustc_codegen_ssa::traits::*;
+                    chunk_to_llval(chunk)
+                }
+                _ => {
+                    // partially uninit, codegen as if it was initialized
```
```suggestion
                    // partially uninit, codegen as if it was initialized (using some arbitrary value for uninit bytes)
```

erikdesjardins commented:
This also needs another perf run, since the latest changes should remove the perf regression

Dylan-DPC commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit b940d411d39cf87dde9d50a08b89fa869dbe1376 with merge f36073c1c18d7834ea40d7e869468ffaaa9fb8a4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b940d411d39cf87dde9d50a08b89fa869dbe1376","merge_sha":"f36073c1c18d7834ea40d7e869468ffaaa9fb8a4"} -->

LingMan commented on compiler/rustc_codegen_llvm/src/consts.rs:
```diff
@@ -11,19 +11,71 @@ use rustc_codegen_ssa::traits::*;
+        // Generating partially-uninit consts inhibits optimizations, so it is disabled by default.
+        // See https://github.com/rust-lang/rust/issues/84565.
+        let allow_partially_uninit =
+            match cx.sess().opts.debugging_opts.partially_uninit_const_threshold {
+                Some(max) => range.len() <= max,
+                None => false,
+            };
```
Unless the formatting ends up being horrible, this looks like an opportunity for `map_or`.

erikdesjardins replied:
It ends up being:

```
        let allow_partially_uninit = cx
            .sess()
            .opts
            .debugging_opts
            .partially_uninit_const_threshold
            .map_or(false, |max| range.len() <= max)
```

I prefer the `match`.

LingMan replied:
Yeah, I guess that qualifies as horrible formatting.

RalfJung replied:
Agreed, `match` is more readable than a higher-order function here. (`map_or` can make code really hard to read IMO.)

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2449131069)
Build commit: f36073c1c18d7834ea40d7e869468ffaaa9fb8a4 (`f36073c1c18d7834ea40d7e869468ffaaa9fb8a4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2449131069"},"merge_sha":"f36073c1c18d7834ea40d7e869468ffaaa9fb8a4"} -->

rust-timer commented:
Queued f36073c1c18d7834ea40d7e869468ffaaa9fb8a4 with parent dc8cb63078c1322cef0ac11c2f8b006f6e310a46, future [comparison URL](https://perf.rust-lang.org/compare.html?start=dc8cb63078c1322cef0ac11c2f8b006f6e310a46&end=f36073c1c18d7834ea40d7e869468ffaaa9fb8a4).

rust-timer commented:
Finished benchmarking try commit (f36073c1c18d7834ea40d7e869468ffaaa9fb8a4): [comparison url](https://perf.rust-lang.org/compare.html?start=dc8cb63078c1322cef0ac11c2f8b006f6e310a46&end=f36073c1c18d7834ea40d7e869468ffaaa9fb8a4).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

erikdesjardins commented:
I managed to reduce it to a test case that fits on godbolt. In both examples, the only difference between the two sources is replacing
`@_ZN3std9panicking4HOOK17hffb77623cb88f088E = internal unnamed_addr global <{ [8 x i8] }> zeroinitializer, align 4`
with
`@_ZN3std9panicking4HOOK17hffb77623cb88f088E = internal unnamed_addr global <{ [4 x i8], [4 x i8] }> <{ [4 x i8] undef, [4 x i8] zeroinitializer }>, align 4`
(and fixing up related GEP indices, types, etc.), and this prevents LLVM from optimizing out the call to the panic hook.

Manually extracted test case (128 lines): https://godbolt.org/z/fv4qf7r7d
Reduced with llvm-reduce (25 lines): https://godbolt.org/z/r6avqsre5

<details>
<summary>llvm-reduce script, for posterity</summary>

```sh
#! /bin/bash

cat $1 > reducednoundef.ll
cat $1 | sed 's/<{ \[8 x i8\] }> zeroinitializer/<{ \[4 x i8\], \[4 x i8\] }> <{ \[4 x i8\] undef, \[4 x i8\] zeroinitializer }>/' |
sed 's/\[8 x i8\]/\[4 x i8\], \[4 x i8\]/g' > reducedundef.ll

./build/x86_64-unknown-linux-gnu/llvm/bin/opt -O3 reducednoundef.ll -S -o reducednoundefopt.ll
./build/x86_64-unknown-linux-gnu/llvm/bin/opt -O3 reducedundef.ll -S -o reducedundefopt.ll

grep -E '_ZN3std9panicking20rust_panic_with_hook17hc745a002b86303d0E\(\)' reducednoundefopt.ll &&
grep -E '_ZN3std9panicking20rust_panic_with_hook17hc745a002b86303d0E\(.+\)' reducedundefopt.ll
```

</details>

oli-obk commented:
Maybe we should report that as a bug in llvm?

nikic commented:
Further reduced: https://godbolt.org/z/3vYqM5865

Dylan-DPC commented:
marking this as blocked because of llvm issue

erikdesjardins commented:
Rebased.

The LLVM issue (tracked in #84565) only occurs with partially-undef consts, so it doesn't block this PR, which doesn't use undef unless the entire const is undef.

@rustbot label: -S-blocked +S-waiting-on-review

LeSeulArtichaut commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 1cf0dac53330610e44a18b111d817a68b76c1497 with merge 7515172ca1fed0db4d6e0cf2031b1a932d6e3078...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1cf0dac53330610e44a18b111d817a68b76c1497","merge_sha":"7515172ca1fed0db4d6e0cf2031b1a932d6e3078"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3270690468)
Build commit: 7515172ca1fed0db4d6e0cf2031b1a932d6e3078 (`7515172ca1fed0db4d6e0cf2031b1a932d6e3078`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3270690468"},"merge_sha":"7515172ca1fed0db4d6e0cf2031b1a932d6e3078"} -->

rust-timer commented:
Queued 7515172ca1fed0db4d6e0cf2031b1a932d6e3078 with parent 57d8747cca741de84918800dca13da0c821e33e0, future [comparison URL](https://perf.rust-lang.org/compare.html?start=57d8747cca741de84918800dca13da0c821e33e0&end=7515172ca1fed0db4d6e0cf2031b1a932d6e3078).

rust-timer commented:
Finished benchmarking try commit (7515172ca1fed0db4d6e0cf2031b1a932d6e3078): [comparison url](https://perf.rust-lang.org/compare.html?start=57d8747cca741de84918800dca13da0c821e33e0&end=7515172ca1fed0db4d6e0cf2031b1a932d6e3078).

**Summary**: This change led to significant regressions 😿 in compiler performance.
- Moderate regression in instruction counts (up to 1.1% on `incr-patched: println` builds of `coercions-opt`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed. 

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

RalfJung commented on compiler/rustc_codegen_llvm/src/consts.rs:
```diff
@@ -11,19 +11,72 @@ use rustc_codegen_ssa::traits::*;
     let pointer_size = dl.pointer_size.bytes() as usize;
 
+    // Note: this function may call `inspect_with_uninit_and_ptr_outside_interpreter`,
+    // so `range` must be within the bounds of `alloc` and not within a relocation.
```
"within a relocation" seems odd -- I think you mean the range must not contain or overlap with a relocation?

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -827,6 +831,49 @@ impl InitMask {
 }
 
+/// Yields [`InitChunk`]s. See [`InitMask::range_as_init_chunks`].
+pub struct InitChunkIter<'a> {
```
There seems to be some conceptual overlap between this and `InitMaskCompressed`... but I am not sure if there is a way to actually share code here.

erikdesjardins replied:
I changed the code that creates `InitMaskCompressed`s to use this...it's not much simpler, but hopefully it'll be faster

RalfJung replied:
Nice! It looks like `compress_uninit_range` is now essentially a more efficient version of `range_as_init_chunks(...).collect::<Vec<_>>`. Can you document it as such? Also it would then make sense to move it *below* `InitChunkIter` in this file.

`InitChunkIter` also seems spread across multiple parts of the file; is it possible to make that code all adjacent to each other?

RalfJung commented:
Hm, any idea why there is a perf regression?

nagisa commented:
I haven't checked the code in detail, but in my memory one of the reasons we've seen in the past is LLVM unable to place something into `.bss` or similar section when the initialiser was a combination of zeroed bytes and uninits. Maybe its that?

RalfJung commented:
If I understand the code correctly, it avoids creating such mixed initializers. They are either fully uninit or fully init.

oli-obk commented:
The regressions are solely in super short perf runs and mostly in `eval_to_allocation_raw` and not really in LLVM. I would say that we're actually seeing the additional overhead due to `is_range_initialized` iterating over `Size` instead of invoking `.as_bytes` first. Somewhere in there LLVM fails to optimize this properly

erikdesjardins commented:
I changed the initialization checks to skip a whole block at once if it's all zeros or ones, and to find the set bit using `trailing_zeros` instead of a loop. It should be faster. ([godbolt](https://godbolt.org/z/35fzv1Wae) for posterity)

erikdesjardins commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -827,10 +820,167 @@ impl InitMask {
+
+    let result = find_bit_fast(init_mask, start, end, is_init);
+
+    debug_assert_eq!(
+        result,
+        find_bit_slow(init_mask, start, end, is_init),
+        "optimized implementation of find_bit is wrong for start={:?} end={:?} is_init={} init_mask={:#?}",
```
Tests + stage 2 build passed with this changed to `assert_eq`, so I'm pretty sure it's correct

RalfJung replied:
That's good! However, I think the code could be made easier to read by adding a few comments. In particular I am missing the big picture here -- there are a lot of small steps and I have no idea why they are all needed. I am not entirely getting this `one_block_past_the_end` business.

Also, helper functions like `search_block` should have comments explaining what exactly they do.

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -563,22 +563,12 @@ impl<Tag, Extra> Allocation<Tag, Extra> {
-                cur = !cur;
-            }
+        for chunk in self.init_mask.range_as_init_chunks(range.start, range.end()) {
+            let len = chunk.range().end.bytes() - chunk.range().start.bytes();
```
This seems to make the assumption that the iterator will always alternate between `InitChunk::Init` and `InitChunk::Uninit`, with the first element matching the value at `range.start`. We should at least clearly document this as part of the contract of `range_as_init_chunks`.

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -831,47 +820,154 @@ impl InitMask {
 /// Yields [`InitChunk`]s. See [`InitMask::range_as_init_chunks`].
 pub struct InitChunkIter<'a> {
     init_mask: &'a InitMask,
+    /// Whether the last chunk was initialized.
```
"previous" would be more clear than "last", I think.

But also, there might not be a previous chunk -- what does this flag mean initially?

erikdesjardins replied:
Hmm this comment is actually backwards--should be the next chunk, not the previous

RalfJung commented:
@bors try @rust-timer queue

bors commented:
:hourglass: Trying commit 4d09eb02063cc48525c2706332a8740bd7bbfa6d with merge 92e1c28b9039ec4a75794c510f2f1a5601b25f3f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"4d09eb02063cc48525c2706332a8740bd7bbfa6d","merge_sha":"92e1c28b9039ec4a75794c510f2f1a5601b25f3f"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3299562456)
Build commit: 92e1c28b9039ec4a75794c510f2f1a5601b25f3f (`92e1c28b9039ec4a75794c510f2f1a5601b25f3f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3299562456"},"merge_sha":"92e1c28b9039ec4a75794c510f2f1a5601b25f3f"} -->

rust-timer commented:
Queued 92e1c28b9039ec4a75794c510f2f1a5601b25f3f with parent d488de82f30fd1dcb0220d57498638596622394e, future [comparison URL](https://perf.rust-lang.org/compare.html?start=d488de82f30fd1dcb0220d57498638596622394e&end=92e1c28b9039ec4a75794c510f2f1a5601b25f3f).

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -827,10 +820,167 @@ impl InitMask {
+        if start_block < one_block_past_the_end {
+            for (&bits, block) in init_mask.blocks[start_block + 1..one_block_past_the_end]
+                .iter()
+                .zip(start_block + 1..)
```
Is the `zip` here to also get the index?

That might be more clear if you used `enumerate`, though maybe that is less performant?
```rust
init_mask.block.iter()
  .enumerate()
  .skip(start_block + 1)
  .take(...)
```

erikdesjardins replied:
Yeah it's written in a specific way to make the main loop simple, codegen is worse with skip/take (and with various other ways of writing it...I spent far too long trying to make a nice looking and fast implementation). I should've added a comment about this, I will now.

rust-timer commented:
Finished benchmarking try commit (92e1c28b9039ec4a75794c510f2f1a5601b25f3f): [comparison url](https://perf.rust-lang.org/compare.html?start=d488de82f30fd1dcb0220d57498638596622394e&end=92e1c28b9039ec4a75794c510f2f1a5601b25f3f).

**Summary**: This change led to significant improvements 🎉 in compiler performance.
- Very large improvement in instruction counts (up to -10.1% on `full` builds of `ctfe-stress-4-opt`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf. 

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

erikdesjardins commented:
Added comments and rearranged some code into (I hope) a more logical order. Let me know if I missed anything

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -704,35 +581,25 @@ pub struct InitMask {
-        m
+    #[inline]
+    fn bit_index(bits: Size) -> (usize, usize) {
+        let bits = bits.bytes();
```
```suggestion
        let bits = bits.bytes(); // BLOCK_SIZE is in bytes.
```

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -825,12 +692,400 @@ impl InitMask {
+                //   start_bit = 3
+                //   is_init = false
+                // Note again that the least significant bit is written first,
+                // which is backwards compared to how we normally write numbers.
```
Why this choice? It also makes `trailing_zeros` somewhat confusing since in this representation it's actually about leading zeros.^^

erikdesjardins replied:
Yeah this really pained me to write...

I think for the other diagrams like:
```
       00000000|00000000
       ^      ^ ^      ^
index: 0      7 8      15
```
it would be more confusing with the MSB first, because then I'd have to write the blocks in reverse, to have the ranges be contiguous:
```
       00000000|00000000
       ^      ^ ^      ^
index: 15     8 7      0
         ^~~~~~~~~~~~^
        end        start
```

And then I made `search_block`'s comments use the same order to be consistent.

But I guess, since I have a disclaimer here anyways, I can just say "Note that, for the examples in this function, the most significant bit is written first, which is backwards compared to the other comments in `find_bit`"

RalfJung replied:
So the real problem here is that the it indexing starts at the LSB... I feel like after a preface explaining that, at least around code that does `trailing_zeros` or `<<`, it'd be better to write bits in the order in which "Rust thinks of them".

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -825,12 +692,400 @@ impl InitMask {
+            //
+            // For (a), the block index of `end_inclusive` is 1, and for (b), it's 0.
+            // This provides the desired behavior of searching blocks 0 and 1 for (a),
+            // and searching only block 0 for (b).
```
```suggestion
            // and searching only block 0 for (b).
            // There is no concern of overflows since we checked for `start >= end` above.
```

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -825,12 +692,400 @@ impl InitMask {
+        let mut ranges = smallvec::SmallVec::<[u64; 1]>::new();
+        let initial = self.init_mask.get(range.start);
+
+        for chunk in self.init_mask.range_as_init_chunks(range.start, range.end()) {
```
```suggestion
        // Here we rely on `range_as_init_chunks` to yield alternating init/uninit chunks.
        for chunk in self.init_mask.range_as_init_chunks(range.start, range.end()) {
```

RalfJung commented:
Those are some very detailed comments, thanks a lot. :) I left some comments.

erikdesjardins commented:
Changed the comments in `search_block` to write the numbers MSB-first (and switched them to use `0b...` binary literal syntax to differentiate them more from the other examples), and made the other requested comment changes

RalfJung commented:
Yeah I saw that, but will need to find the time to read over all the Miri changes again. I'm moving to a different continent next week so things are a bit crazy.^^

bors commented:
:umbrella: The latest upstream changes (presumably #88242) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -825,12 +695,402 @@ impl InitMask {
+            // In both cases, the block index of `end` is 1.
+            // But we do want to search block 1 in (a), and we don't in (b).
+            //
+            // If we subtract 1 from both end positions to make them inclusive:
```
```suggestion
            // We subtract 1 from both end positions to make them inclusive:
```

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -825,12 +695,402 @@ impl InitMask {
+    pub fn range_as_init_chunks(&self, start: Size, end: Size) -> InitChunkIter<'_> {
+        assert!(end <= self.len);
+
+        let is_init = if start < end { self.get(start) } else { false };
```
Does the `false` here have any consequences? If not, maybe add a comment saying that this is basically arbitrary.

RalfJung commented on compiler/rustc_middle/src/mir/interpret/allocation.rs:
```diff
@@ -825,12 +695,402 @@ impl InitMask {
+        // where each element toggles the state.
+
+        let mut ranges = smallvec::SmallVec::<[u64; 1]>::new();
+        let initial = self.init_mask.get(range.start);
```
I am a bit worried that getting `initial` separately from `range_as_init_chunks` can lead to a bug where the iterator is "inverted" from the state we are expecting... is there any way to avoid this, using just `range_as_init_chunks` as the sole source of information here?

RalfJung commented:
The changes in `interpret` look good to me, modulo the comments I just left.

I hope @oli-obk or someone else can do the review of the other changes.

the8472 commented:
While the perf results show improvements in instructions counts this looks like a significant (non-noise) [regression in wall times](https://perf.rust-lang.org/compare.html?start=3b3ce374d203445eb1d0dce50f4211f4aceb7db6&end=20997f6ad81721542e9ef97bb2f58190903a34d8&stat=wall-time).

The jump is noticable [in the timelines](https://perf.rust-lang.org/index.html?start=2021-08-26&end=2021-08-31&kind=percentfromfirst&stat=wall-time) of `encodings-opt` or `await-call-tree-opt`

Labeling as regression so that the perf team can have a look at it.

oli-obk commented:
the await call tree opt change is in #88308 not in this PR

there are a few other noticeable regressions in the timelines though, yes.

Most of the non-incremental regressions I looked at seem to point to LLVM spending more time, so this may either end up with better optimized code or just make it harder for LLVM to figure things out.

the8472 commented:
`hello-world-check` regressed too (especially incr-unchanged), that shouldn't involve llvm.

erikdesjardins commented:
I tried to reproduce this locally but was unable to. I don't see a regression in wall-time, nor when running under cachegrind. Cachegrind actually shows a slight improvement for `encoding-opt`.

(For posterity, using e.g. `cargo +nightly run --release --bin collector -- diff_local cachegrind --include helloworld --builds Opt --runs IncrUnchanged +3b3ce374d203445eb1d0dce50f4211f4aceb7db6 +20997f6ad81721542e9ef97bb2f58190903a34d8`, with a patched version of the collector that sets `--cache-sim=yes --branch-sim=yes` and runs `rustfilt`.)

So, I decided to look at the [timeline](https://perf.rust-lang.org/index.html?start=2021-07-25&end=2021-10-01&kind=percentfromfirst&stat=wall-time) now that more time has passed and...it seems kinda weird.

Zoomed out, it's not that unusual. This PR introduced a regression and then a later change fixed it--right?

![image](https://user-images.githubusercontent.com/7673145/135564503-7b01b29a-f4e2-4272-ac02-489fb05b3966.png)

But if we zoom in on the point where wall-time returns to normal, it doesn't make sense:

![image](https://user-images.githubusercontent.com/7673145/135564835-f2485faf-ba55-40ca-86ac-c7017491d8ef.png)

First, it bounces between slow and fast before settling on fast. Which seems implausible--it's not likely that we introduced a perf improvement, then cancelled it out in the next commit, then reintroduced it three commits later.

Second, the two commits responsible for those two improvements don't make sense:
- 38e576423dc56a006ccf0b72e106f09bdea6efc2: This is a perf improvement, but it just avoids executing code. So for a >10% decrease in wall-time you'd expect a significant decrease in instruction count, but `helloworld-opt` only has [-0.02%](https://perf.rust-lang.org/compare.html?start=78a46efff06558674b51c10d8d81758285746ab5&end=38e576423dc56a006ccf0b72e106f09bdea6efc2).
- 9dd4ce80fb01d1ff5cb5002f08b7b3847b59e664: This is a cargo update and none of the cargo changes seem related to perf.

(All `helloworld` and `encoding` benchmarks exhibit this behavior. And probably other ones but I didn't check.)

I'm not sure what to make of this. Some intermittent issue with the perf collector that makes wall-time measurements longer?

---

Actually, while typing this up, I think I figured out what it is.

See this comment about the cpu governor being reset on the perf machine: https://github.com/rust-lang/rust/pull/88956#issuecomment-922277439

#88308 was merged immediately after this PR. It too had some (unexplainable) wall-time regressions, in `await-call-tree`, `deeply-nested-closures`, and `deeply-nested-async`.

Similarly, if you look at the [timeline](https://perf.rust-lang.org/index.html?start=2021-07-25&end=2021-10-01&kind=percentfromfirst&stat=wall-time) again:

![image](https://user-images.githubusercontent.com/7673145/135571889-61888905-426b-4bd0-9b65-de00140b51a0.png)

And zoom in:

![image](https://user-images.githubusercontent.com/7673145/135572021-95fa8fb7-6277-493c-a7ce-6936ee8898b7.png)

The improvements are due to 38e576423dc56a006ccf0b72e106f09bdea6efc2 (again), and a58db2e4ddbb2e0fa91c53c4901b52a7ef71ef2d (merged right after 9dd4ce80fb01d1ff5cb5002f08b7b3847b59e664).

So I think this is what happened:

- This PR is merged (20997f6ad81721542e9ef97bb2f58190903a34d8). Halfway though the perf run, the cpu governor gets reset, regressing the wall-time of remaining benchmarks.

- #88308 is merged right after (4b9f4b221b92193c7e95b1beb502c6eb32c3b613). It "regresses" the benchmarks that the previous run had already completed before the governor changed.

<time passes>

- 9dd4ce80fb01d1ff5cb5002f08b7b3847b59e664 hits the perf collector. Halfway through, the cpu governor gets fixed, undoing the regression in some benchmarks.

- a58db2e4ddbb2e0fa91c53c4901b52a7ef71ef2d comes right after, undoing it in the rest.

And judging from the timestamp on [this comment](https://github.com/rust-lang/rust/pull/88934#issuecomment-924039469), 38e576423dc56a006ccf0b72e106f09bdea6efc2 was missed initially and manually re-enqueued later. This would explain the downward spike that appears everywhere--although it was committed earlier, its perf run occurred later, after the governor was fixed.

[](https://user-images.githubusercontent.com/7673145/135572706-a31559f8-6367-4c20-8dde-c7d340975dfa.png "the spike")

the8472 commented:
That seems like a reasonable explanation, marking as triaged.
